### PR TITLE
Implement AudioPacketInterceptor

### PR DESCRIPTION
**Pull request by mentlerd**
Implemented a new interface for AudioPacketInterceptor.

This object can be registered just like an AudioReceiveHandler, but it provides a much more low level access to audio packets than the standard handleUserAudio callback.

Also allows discarding AudioPackets, effectively hiding them from the JDA backend.

This PR implements issue #418

### CODE REVIEW

**Comment by mentlerd** [comment]
This pull request belonds to issue #425

File: src/main/java/net/dv8tion/jda/core/audio/AudioConnection.java [hunk 1]
```diff
@@ -258,28 +266,33 @@ else if (sendHandler == null && sendSystem != null)
 
     private synchronized void setupReceiveSystem()
     {
```
Comment (mentlerd) [inline src/main/java/net/dv8tion/jda/core/audio/AudioConnection.java]: The current implementation initializes the receive thread even if there is only an AudioPacketInterceptor, but no ReceiveHandler.

I have tried to cleanup this setup routine as much as I could, but I would appreciate a second pair of eyes checking on it.

Comment (kantenkugel) [inline src/main/java/net/dv8tion/jda/core/audio/AudioConnection.java]: With your changes, if there is a "consumer" but the socket is either null or closed, nothing happens.
In the old implementation, if the socket was null/closed (no matter if "consumer" is set, there was cleanup (your else branch)

Comment (mentlerd) [inline src/main/java/net/dv8tion/jda/core/audio/AudioConnection.java]: Thanks, I did not spot that!

I think it does not make a difference in this case tough, the combined audio thread only gets setup when the udpSocket is not null, so there is nothing to clean up.

**Comment by MinnDevelopment** [comment]
This PR is on hold until voice resume has been merged. On a side note, you need to update your code style to fit into JDA, the wiki explains this.

**Comment by mentlerd** [comment]
I have fixed the teardown and style issues

**Comment by MinnDevelopment** [comment]
I'm assigning this to @DV8FromTheWorld so he can decide how we should implement this.

**Comment by MinnDevelopment** [comment]
Closing this, so @DV8FromTheWorld can use the issue to decide on an implementation. This PR is not satisfying.

**Pull request closed without merging**